src/modules/schnorr_adaptor/main_impl.h

```diff
@@ -0,0 +1,286 @@
+    secp256k1_fe_normalize_var(&r0.y);
+    if (secp256k1_fe_is_odd(&r0.y)) {
+        secp256k1_scalar_negate(&k, &k);
+    }
+
+    ret &= !!secp256k1_eckey_pubkey_serialize(&r0, sig65, &size, 1);
+
+    secp256k1_schnorrsig_challenge(&e, &sig65[1], msg32, msglen, pk_buf);
+    secp256k1_scalar_mul(&e, &e, &sk);
+    /* k + e * d */
+    secp256k1_scalar_add(&e, &e, &k);
+    secp256k1_scalar_get_b32(&sig65[33], &e);
+
+    secp256k1_memczero(sig65, 65, !ret);
+    secp256k1_scalar_clear(&k);
+    secp256k1_scalar_clear(&sk);
+    memset(seckey, 0, sizeof(seckey));
+
+    return ret;
+}
+
+int secp256k1_schnorr_adaptor_presign(const secp256k1_context *ctx, unsigned char *sig65, const unsigned char *msg32, const secp256k1_keypair *keypair, const unsigned char *t33, const unsigned char *aux_rand32) {
+    /* We cast away const from the passed aux_rand32 argument since we know the default nonce function does not modify it. */
+    return secp256k1_schnorr_adaptor_presign_internal(ctx, sig65, msg32, keypair, secp256k1_adaptor_nonce_function_bip340, t33, (unsigned char*)aux_rand32);
+}
+
+int secp256k1_schnorr_adaptor_extract_t(const secp256k1_context *ctx, unsigned char *t33, const unsigned char *sig65, const unsigned char *msg32, const secp256k1_xonly_pubkey *pubkey) {
+    secp256k1_scalar s0;
+    secp256k1_scalar e;
+    secp256k1_gej rj;
+    secp256k1_ge r;
+    secp256k1_ge pk;
+    secp256k1_gej pkj;
+    secp256k1_ge r0;
+    secp256k1_ge t;
+    secp256k1_gej tj;
+    secp256k1_xonly_pubkey pkr0;
+    unsigned char buf[32];
+    size_t size = 33;
+    size_t msglen = 32;
+    int overflow;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(t33 != NULL);
+    ARG_CHECK(sig65 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkey != NULL);
+
+    /* P */
+    ret &= !!secp256k1_xonly_pubkey_load(ctx, &pk, pubkey);
+
+    /* s0 */
+    secp256k1_scalar_set_b32(&s0, &sig65[33], &overflow);
+    ret &= !overflow;
+
+    /* R0 */
+    ret &= !!secp256k1_xonly_pubkey_parse(ctx, &pkr0, &sig65[1]);
+    ret &= !!secp256k1_xonly_pubkey_load(ctx, &r0, &pkr0);
+
+    /* Compute e */
+    secp256k1_fe_get_b32(buf, &pk.x);
+    secp256k1_schnorrsig_challenge(&e, &sig65[1], msg32, msglen, buf);
+
+    /* Compute rj = s0*G + (-e) * pkj */
+    secp256k1_scalar_negate(&e, &e);
+    secp256k1_gej_set_ge(&pkj, &pk);
+    secp256k1_ecmult(&rj, &pkj, &e, &s0);
+
+    /* R */
+    secp256k1_ge_set_gej_var(&r, &rj);
+    ret &= !secp256k1_ge_is_infinity(&r);
+
+    /* T = R0 + (- R) */
+    secp256k1_gej_neg(&rj, &rj);
+    secp256k1_gej_add_ge_var(&tj, &rj, &r0, NULL);
+    if (sig65[0] == SECP256K1_TAG_PUBKEY_EVEN) {
+        ;
+    } else if (sig65[0] == SECP256K1_TAG_PUBKEY_ODD) {
+        secp256k1_gej_neg(&tj, &tj);
+    } else {
+        ret = 0;
+    }
+    secp256k1_ge_set_gej(&t, &tj);
+    ret &= !!secp256k1_eckey_pubkey_serialize(&t, t33, &size, 1);
+
+    secp256k1_memczero(t33, 33, !ret);
```

Comment:
```suggestion
    memset(t33, 0, 33);
```
The function can be variable time. We probably need to perform this at the start of the function if we plan to `return 0` in case of failures.

Reply:
Sounds good

Reply:
I used secp256k1_memczero here because if ret == 0, t33 must be set 0.

Reply:
Oh, okay. Then, you can use: `if(!ret) memset(t33, 0, 33)`, when you make this func variable time.